File: services/ExampleService.ts
import { ServiceClient } from "..";
import { service } from "../util";

declare global {
    namespace services {
        const ExampleService: ServiceClient<ExampleService>;
    }
}

export type HelloRequest = {
    name: string;
};

export type HelloReply = {
    message: string;
};

@service("services.ExampleService")
export default class ExampleService {
    async sayHello(req: HelloRequest): Promise<HelloReply> {
        return await Promise.resolve({ message: "Hello, " + req.name });
    }
}

File: services/ExampleService.go
package services

import (
	"context"

	"github.com/ayonli/ngrpc"
	"github.com/ayonli/ngrpc/services/proto"
	"google.golang.org/grpc"
)

type ExampleService struct {
	proto.UnimplementedExampleServiceServer
}

func (self *ExampleService) Serve(s grpc.ServiceRegistrar) {
	proto.RegisterExampleServiceServer(s, self)
}

func (self *ExampleService) Connect(cc grpc.ClientConnInterface) proto.ExampleServiceClient {
	return proto.NewExampleServiceClient(cc)
}

func (self *ExampleService) GetClient(route string) (proto.ExampleServiceClient, error) {
	return ngrpc.GetServiceClient(self, route)
}

func (self *ExampleService) SayHello(ctx context.Context, req *proto.HelloRequest) (*proto.HelloReply, error) {
	return &proto.HelloReply{Message: "Hello, " + req.Name}, nil
}

func init() {
	ngrpc.Use(&ExampleService{})
}
